### ISSUE TITLE
The sidecar container does not work well with istio-proxy sidecar container

### ISSUE THREAD

**Issue by songjiaxun** [GoogleCloudPlatform/gcs-fuse-csi-driver#53]
# Symptom

If the gcsfuse sidecar container starts before the istio-proxy sidecar container, the gcsfuse will fail with the following error:

> mountWithArgs: failed to open connection - getConnWithRetry: get token source: DefaultTokenSource: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.

# Root cause

The same as https://github.com/GoogleCloudPlatform/gcs-fuse-csi-driver/issues/46#issue-1812501442.

# Solution

**This issue is tracked on [GitHub](https://github.com/GoogleCloudPlatform/gcs-fuse-csi-driver/issues/20), and we are waiting for the [Kubernetes sidecar container feature](https://github.com/kubernetes/kubernetes/pull/116429) available on GKE to ultimately solve this issue.**

**Comment by jpmnteiro** [comment]
Ran into a similar problem (on k8s 1.29 w/ the new sidecar container feature - running on GKE). The initContainers injected by istio fail to start/move to ready and the pod is left in PodInitialisating.

The initContainers:

```
  initContainers:
  - args:
    - istio-iptables
    - -p
    - "15001"
    - -z
    - "15006"
    - -u
    - "1337"
    - -m
    - REDIRECT
    - -i
    - '*'
    - -x
    - ""
    - -b
    - '*'
    - -d
    - 15090,15021,15020
    - --log_output_level=default:info
    - --run-validation
    - --skip-rule-apply
    ...
    image: gcr.io/gke-release/asm/proxyv2:1.18.7-asm.4
    imagePullPolicy: IfNotPresent
    name: istio-validation
    resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: 100m
        memory: 128Mi
    restartPolicy: Always
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    ...
  - args:
    - --v=5
    env:
    - name: NATIVE_SIDECAR
      value: "TRUE"
    image: gke.gcr.io/gcs-fuse-csi-driver-sidecar-mounter:v1.2.0-gke.0@sha256:31880114306b1fb5d9e365ae7d4771815ea04eb56f0464a514a810df9470f88f
    imagePullPolicy: IfNotPresent
    name: gke-gcsfuse-sidecar
    resources:
      requests:
        cpu: 250m
        ephemeral-storage: 5Gi
        memory: 256Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      readOnlyRootFilesystem: true
      runAsGroup: 65534
      runAsNonRoot: true
      runAsUser: 65534
      seccompProfile:
        type: RuntimeDefault
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /gcsfuse-tmp
      name: gke-gcsfuse-tmp
    - mountPath: /gcsfuse-buffer
      name: gke-gcsfuse-buffer
    - mountPath: /gcsfuse-cache
      name: gke-gcsfuse-cache
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-x7xtn
      readOnly: true
```

**Comment by mgnisia** [comment]
also came across your issue you described here today, I used the following pod annotation to make it work: 

```
proxy.istio.io/config: '{ "holdApplicationUntilProxyStarts": true }'
```

**Comment by songjiaxun** [comment]
This is fixed in GKE 1.29 and above clusters.

To avoid the errors with ASM or istio, you need to:
1. Make sure all the nodes are upgraded to 1.29+.
2. Add pod annotation `proxy.istio.io/config: '{ "holdApplicationUntilProxyStarts": true }'`.
3. Add pod annotation `traffic.sidecar.istio.io/excludeOutboundIPRanges: 169.254.169.254/32`.
